File: algorithms-week9/bfs/cpp/bfs.cpp
//
// Created by Sam on 9/12/2018.
//

#include <iostream>
#include <vector>
#include <queue>

using std::vector;
using std::queue;

static void bfs(vector<vector<int>>& adj, vector<int>& dist, queue<int>& q) {
  while (!q.empty()) {
    int v = q.front();
    q.pop();

    for (int& w : adj[v]) {
      if (dist[w] == -1) {
        q.push(w);
        dist[w] = dist[v] + 1;
      }
    }
  }
}

int distance(vector<vector<int> > &adj, int s, int t) {
  size_t n = adj.size();
  vector<int> dist(n, -1);
  queue<int> q;

  dist[s] = 0;
  q.push(s);

  bfs(adj, dist, q);
  return dist[t];
}

int main() {
  int n, m;
  std::cin >> n >> m;
  vector<vector<int> > adj(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int x, y;
    std::cin >> x >> y;
    adj[x - 1].push_back(y - 1);
    adj[y - 1].push_back(x - 1);
  }
  int s, t;
  std::cin >> s >> t;
  s--, t--;
  std::cout << distance(adj, s, t);
}

File: algorithms-week9/bfs/python/bfs.py
#Uses python3

import sys
import queue

def distance(adj, s, t):
    n = len(adj)
    dist = [-1] * n
    q = queue.Queue(n)

    dist[s] = 0
    q.put(s)

    bfs(adj, dist, q)
    return dist[t]

def bfs(adj, dist, q):
    while not q.empty():
        v = q.get()

        for w in adj[v]:
            if dist[w] == -1:
                q.put(w)
                dist[w] = dist[v] + 1

    return

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
        adj[b - 1].append(a - 1)
    s, t = data[2 * m] - 1, data[2 * m + 1] - 1
    print(distance(adj, s, t))
